language_server/src/lsp/lsp_server.cpp

```diff
@@ -247,59 +252,108 @@ std::string replace_empty_by_space(std::string s)
     else
         return s;
 }
+
+json create_diag_json(const parser_library::range& r,
+    const char* code,
+    const char* source,
+    const char* message,
+    std::optional<json> diag_related_info,
+    parser_library::diagnostic_severity severity,
+    parser_library::diagnostic_tag tags)
+{
+    json one_json {
+        { "range", feature::range_to_json(r) },
+        { "code", code },
+        { "source", source },
+        { "message", replace_empty_by_space(message) },
+    };
+
+    if (diag_related_info.has_value())
+        one_json["relatedInformation"] = std::move(*diag_related_info);
+
+    if (severity != parser_library::diagnostic_severity::unspecified)
+        one_json["severity"] = static_cast<int>(severity);
+
+    if (tags != parser_library::diagnostic_tag::none)
+    {
+        auto& j_tags = one_json["tags"] = json::array();
+        if (static_cast<int>(tags) & static_cast<int>(parser_library::diagnostic_tag::unnecessary))
+            j_tags.push_back(1);
+        if (static_cast<int>(tags) & static_cast<int>(parser_library::diagnostic_tag::deprecated))
+            j_tags.push_back(2);
+    }
+
+    return one_json;
+}
 } // namespace
 
-void server::consume_diagnostics(parser_library::diagnostic_list diagnostics)
+void server::consume_diagnostics(
+    parser_library::diagnostic_list diagnostics, parser_library::fade_message_list fade_messages)
 {
     // map of all diagnostics that came from the server
-    std::map<std::string, std::vector<parser_library::diagnostic>> diags;
+    std::map<std::string,
```

Comment:
I don't quite remember why this map was needed at all... it seems like building the json directly has not downside...

Reply:
I guess, it is about "consolidating" the diagnostic messages into a single vector which is later processed for each file?
Therefore, there is only one json for a single file. 

I'm not sure if that's still a requirement by VSCode but I can try it out.

Reply:
true, but storing json::array_t value directly seems easier